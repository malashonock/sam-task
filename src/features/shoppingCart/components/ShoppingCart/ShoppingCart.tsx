import { useContext } from 'react';
import ShoppingCartContext, { ShoppingCartState } from '../../model/ShoppingCartContext';
import CartItemsCounter from '../CartItemsCounter/CartItemsCounter';

export default function ShoppingCart(): JSX.Element {
  const {
    shoppingCart: { itemsCount },
  }: ShoppingCartState = useContext(ShoppingCartContext) as ShoppingCartState;

  return (
    <i className="bi bi-cart4 position-relative">
      {itemsCount > 0 && <CartItemsCounter count={itemsCount} />}
    </i>
  );
}
