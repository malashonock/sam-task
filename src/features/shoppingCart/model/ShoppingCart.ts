import Guitar, { areGuitarsEqual } from 'features/guitars/model/Guitar';
import ShoppingCartLineItem from './ShoppingCartLineItem';

export interface IShoppingCart {
  lineItems: ShoppingCartLineItem[];
  itemsCount: number;
  pendingOverflow: boolean;
}

export default class ShoppingCart implements IShoppingCart {
  static MAX_ITEMS_COUNT: number = 20;

  constructor(
    public lineItems: ShoppingCartLineItem[] = [],
    public itemsCount: number = 0,
    public pendingOverflow: boolean = false
  ) {
    this.lineItems = lineItems;
    this.itemsCount = itemsCount;
    this.pendingOverflow = pendingOverflow;
  }

  findLineItemIndex(itemToMatch: Guitar): number {
    return this.lineItems
      .map((lineItem: ShoppingCartLineItem): Guitar => lineItem.sku)
      .findIndex((itemInCart: Guitar): boolean => areGuitarsEqual(itemInCart, itemToMatch));
  }

  hasItem(item: Guitar): boolean {
    return this.findLineItemIndex(item) >= 0;
  }

  addItem(itemToAdd: Guitar): void {
    if (this.itemsCount === ShoppingCart.MAX_ITEMS_COUNT) {
      // signal overflow attempt
      this.pendingOverflow = true;
      return;
    }

    const lineItemToIncrementIndex = this.findLineItemIndex(itemToAdd);

    // increment items count
    this.itemsCount++;

    // such line item already exists?
    if (lineItemToIncrementIndex >= 0) {
      // increment matched line item
      const lineItemToIncrement = this.lineItems[lineItemToIncrementIndex];
      lineItemToIncrement.quantity++;
    } else {
      // add new item
      this.lineItems.push({
        sku: itemToAdd,
        quantity: 1,
      });
    }
  }

  removeItem(itemToRemove: Guitar): void {
    const lineItemToDecrementIndex = this.findLineItemIndex(itemToRemove);

    if (lineItemToDecrementIndex < 0) {
      // no such item, nothing to remove
      return;
    }

    // in any case, decrement items count
    this.itemsCount--;

    const lineItemToDecrementQuantity = this.lineItems[lineItemToDecrementIndex].quantity;

    // should line item remain in cart?
    if (lineItemToDecrementQuantity > 1) {
      // decrement matched line item
      this.lineItems[lineItemToDecrementIndex].quantity--;
    } else {
      // remove matched line item
      this.lineItems.splice(lineItemToDecrementIndex, 1);
    }
  }

  removeLineItem(itemToRemove: Guitar): void {
    const lineItemToDecrementIndex = this.findLineItemIndex(itemToRemove);

    if (lineItemToDecrementIndex < 0) {
      // no such item, nothing to remove
      return;
    }

    const lineItemToDecrementQuantity = this.lineItems[lineItemToDecrementIndex].quantity;

    // decrease item count by the quantity of line item being removed
    this.itemsCount -= lineItemToDecrementQuantity;

    // remove line item
    this.lineItems.splice(lineItemToDecrementIndex, 1);
  }

  clone(): ShoppingCart {
    return new ShoppingCart([...this.lineItems], this.itemsCount, this.pendingOverflow);
  }
}
