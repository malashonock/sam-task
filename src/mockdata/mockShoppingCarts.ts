import ShoppingCart from '../features/shoppingCart/model/ShoppingCart';
import mockGuitars from './mockGuitars';

const mockShoppingCarts: { [name: string]: ShoppingCart } = {
  empty: {
    guitars: [],
    pendingOverflow: false,
  },

  oneGuitarNoOverflow: {
    guitars: mockGuitars.oneGuitar,
    pendingOverflow: false,
  },

  oneGuitarPendingOverflow: {
    guitars: mockGuitars.oneGuitar,
    pendingOverflow: true,
  },
};

export default mockShoppingCarts;
