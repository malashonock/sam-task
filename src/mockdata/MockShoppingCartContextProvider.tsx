import { PropsWithChildren, useReducer } from 'react';
import ShoppingCart from '../features/shoppingCart/model/ShoppingCart';
import ShoppingCartContext from '../features/shoppingCart/model/ShoppingCartContext';
import { shoppingCartReducer } from '../features/shoppingCart/model/ShoppingCartReducer';

export default function MockShoppingCartContextProvider({
  mockShoppingCart,
  children,
}: { mockShoppingCart: ShoppingCart } & PropsWithChildren): JSX.Element {
  const [shoppingCart, dispatchShoppingCartAction] = useReducer(
    shoppingCartReducer,
    mockShoppingCart
  );

  return (
    <ShoppingCartContext.Provider value={{ shoppingCart, dispatchShoppingCartAction }}>
      {children}
    </ShoppingCartContext.Provider>
  );
}
