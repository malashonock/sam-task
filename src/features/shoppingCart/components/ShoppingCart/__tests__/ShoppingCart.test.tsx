import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import MockShoppingCartContextProvider from '../../../../../mockdata/MockShoppingCartContextProvider';
import mockShoppingCarts from '../../../../../mockdata/mockShoppingCarts';
import ShoppingCart from '../ShoppingCart';

describe('ShoppingCart component', () => {
  it('matches snapshot', () => {
    const shoppingCart = renderer.create(
      <MockShoppingCartContextProvider mockShoppingCart={mockShoppingCarts.empty}>
        <ShoppingCart />
      </MockShoppingCartContextProvider>
    );

    expect(shoppingCart).toMatchSnapshot();
  });

  describe('given no guitars in shopping cart', () => {
    it('renders no CartItemsCounter', () => {
      const { container } = render(
        <MockShoppingCartContextProvider mockShoppingCart={mockShoppingCarts.empty}>
          <ShoppingCart />
        </MockShoppingCartContextProvider>
      );

      const cartItemsCounter = container.querySelector('.badge');
      expect(cartItemsCounter).toBeNull();
    });
  });

  describe('given some guitars in shopping cart', () => {
    it('renders CartItemsCounter', () => {
      const { container } = render(
        <MockShoppingCartContextProvider mockShoppingCart={mockShoppingCarts.oneGuitarNoOverflow}>
          <ShoppingCart />
        </MockShoppingCartContextProvider>
      );

      const cartItemsCounter = container.querySelector('.badge');
      expect(cartItemsCounter).not.toBeNull();
    });
  });
});
