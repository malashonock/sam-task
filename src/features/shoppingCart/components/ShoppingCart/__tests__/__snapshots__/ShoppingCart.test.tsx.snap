// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart component matches snapshot 1`] = `
<i
  className="bi bi-cart4 position-relative fs-2"
/>
`;
